### ISSUE TITLE
Bug: null as default option is failing in applyDefault

### ISSUE THREAD

**Issue by gajus** [typescript-eslint/typescript-eslint#6439]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Repro Code

```typescript
/**
 * @file Rule to ensure that filenames match the exports of the file
 * @author Stefan Lau
 * @see https://github.com/selaux/eslint-plugin-filenames/blob/32fc70dd7572211d1e5b97e06ec7a005c77fe8d4/lib/rules/match-exported.js
 */

import path from 'node:path';
import { camelCase, kebabCase, upperFirst, snakeCase } from 'lodash';
import {
  createRule,
  isIgnoredFilename,
  getExportedName,
  isIndexFile,
  parseFilename,
} from '../utilities';

const transformMap = {
  camel: camelCase,
  kebab: kebabCase,
  pascal(name) {
    return upperFirst(camelCase(name));
  },
  snake: snakeCase,
};

const getStringToCheckAgainstExport = (parsed, replacePattern?: RegExp) => {
  const directoryArray = parsed.dir.split(path.sep);
  const lastDirectory = directoryArray[directoryArray.length - 1];

  if (isIndexFile(parsed)) {
    return lastDirectory;
  }

  return replacePattern ? parsed.name.replace(replacePattern, '') : parsed.name;
};

const getTransformsFromOptions = (option) => {
  const usedTransforms = option?.push ? option : [option];

  return usedTransforms.map((name) => {
    return transformMap[name];
  });
};

const transform = (exportedName, transforms) => {
  return transforms.map((format) => {
    return format ? format(exportedName) : exportedName;
  });
};

const anyMatch = (expectedExport, transformedNames) => {
  return transformedNames.includes(expectedExport);
};

const getWhatToMatchMessage = (transforms) => {
  if (transforms.length === 1 && !transforms[0]) {
    return 'the exported name';
  }

  if (transforms.length > 1) {
    return 'any of the exported and transformed names';
  }

  return 'the exported and transformed name';
};

type Options = [string[] | string | null, string | null, boolean];

type MessageIds = 'indexFile' | 'regularFile';

export default createRule<Options, MessageIds>({
  create: (context, options) => {
    return {
      Program(node) {
        const transforms = getTransformsFromOptions(options[0]);
        const replacePattern = options[1]
          ? new RegExp(options[1], 'u')
          : undefined;
        const filename = context.getFilename();
        const absoluteFilename = path.resolve(filename);
        const parsed = parseFilename(absoluteFilename);
        const shouldIgnore = isIgnoredFilename(filename);
        const exportedName = getExportedName(node, options[2]);
        const isExporting = Boolean(exportedName);
        const expectedExport = getStringToCheckAgainstExport(
          parsed,
          replacePattern,
        );
        const transformedNames = transform(exportedName, transforms);
        const everythingIsIndex =
          exportedName === 'index' && parsed.name === 'index';
        const matchesExported =
          anyMatch(expectedExport, transformedNames) || everythingIsIndex;

        if (shouldIgnore || !isExporting || matchesExported) {
          return;
        }

        context.report({
          data: {
            expectedExport,
            exportName: transformedNames.join("', '"),
            extension: parsed.ext,
            name: parsed.base,
            whatToMatch: getWhatToMatchMessage(transforms),
          },
          messageId: isIndexFile(parsed) ? 'indexFile' : 'regularFile',
          node,
        });
      },
    };
  },
  defaultOptions: [null, null, false],
  meta: {
    docs: {
      description:
        'Match the file name against the default exported value in the module.',
      recommended: 'warn',
    },
    messages: {
      indexFile:
        "The directory '{{expectedExport}}' must be named '{{exportName}}', after the exported value of its index file.",
      regularFile:
        "Filename '{{expectedExport}}' must match {{whatToMatch}} '{{exportName}}'.",
    },
    schema: [
      {
        oneOf: [
          {
            items: {
              type: 'string',
            },
            type: 'array',
          },
          {
            type: 'string',
          },
          {
            type: 'null',
          },
        ],
      },
      {
        oneOf: [
          {
            type: 'string',
          },
          {
            type: 'null',
          },
        ],
      },
      {
        type: 'boolean',
      },
    ],
    type: 'suggestion',
  },
  name: 'filename-match-exported',
});
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

When invoked with options `[null, null, true],`, this rule is failing with a rather generic error:

```
TypeError: Error while loading rule 'filename-match-exported': Cannot convert undefined or null to object
Occurred while linting /some/dir/foo.js
      at Function.keys (<anonymous>)
      at deepMerge (node_modules/.pnpm/@typescript-eslint+utils@5.51.0_4vsywjlpuriuw3tl5oq6zy5a64/node_modules/@typescript-eslint/utils/src/eslint-utils/deepMerge.ts:26:31)
...
```

After digging into the stack trace, I found that the error originates in `applyDefault` function, and the actual error is:

```
TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at deepMerge (/Users/gajus/Documents/dev/gajus/eslint-plugin-canonical/node_modules/.pnpm/@typescript-eslint+utils@5.51.0_4vsywjlpuriuw3tl5oq6zy5a64/node_modules/@typescript-eslint/utils/src/eslint-utils/deepMerge.ts:26:31)
```

### Actual Result

Expected user options to override defaults.

### Versions

N/A

**Comment by bradzacher** [comment]
This is the first time we've ever seen someone use `null` as a default at the top-level.
Generally we also recommend using objects for config instead of the array form because the array form can be pretty confusing inside the config because it contains no context for what each element might be exactly.

EG this style of config is pretty cryptic and would require anyone to go look up the rule's docs to understand what the config might actually mean.
```
rules: {
  'custom/rule': ['error', ['config1', 'config2', 'config3'], null, false]],
}
```
Additionally the array style requires that config options be passed in a specific order, which does make things hard when a user wants to ignore specific config options but pass others.

OTOH using an object is self-documenting to some degree.
```
rules: {
  'custom/rule': ['error', {
    clearConfigName: ['config1', 'config2', 'config3'],
    otherConfigName: null, // or... could be omitted
    booleanFlag: false,
  }],
}
```

---

To be clear - not saying this isn't something we should fix, just explaining why nobody has encountered this before.

**Comment by gajus** [comment]
For context, I am migrating various legacy ESLint rules to https://github.com/gajus/eslint-plugin-canonical

In this case, `filename-*` rules were using this pattern. Similar to what you've suggested, I just refactored them to use object, but still filled the bug for your awareness.
